crates/red_knot_python_semantic/src/types.rs

```diff
@@ -381,16 +382,68 @@ mod tests {
         )
         .expect("Valid search path settings");
 
+        db
+    }
+
+    fn assert_diagnostic_messages(diagnostics: &TypeCheckDiagnostics, expected: &[&str]) {
+        let messages: Vec<&str> = diagnostics
+            .iter()
+            .map(|diagnostic| diagnostic.message())
+            .collect();
+        assert_eq!(&messages, expected);
+    }
+
+    #[test]
+    fn check_types() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_file("src/foo.py", "import bar\n")
+            .context("Failed to write foo.py")?;
+
         let foo = system_path_to_file(&db, "src/foo.py").context("Failed to resolve foo.py")?;
 
         let diagnostics = super::check_types(&db, foo);
+        assert_diagnostic_messages(&diagnostics, &["Import 'bar' could not be resolved."]);
+
+        Ok(())
+    }
 
-        assert_eq!(diagnostics.len(), 1);
-        assert_eq!(
-            diagnostics[0].message(),
-            "Import 'bar' could not be resolved."
+    #[test]
+    fn unresolved_import() {
+        let mut db = setup_db();
+
+        db.write_files([
+            ("/src/a.py", "import foo as foo"),
+            ("/src/b.py", "from a import foo"),
+        ])
+        .unwrap();
+
+        let a_file = system_path_to_file(&db, "/src/a.py").expect("Expected `a.py` to exist!");
+        let a_file_diagnostics = super::check_types(&db, a_file);
+        assert_diagnostic_messages(
+            &a_file_diagnostics,
+            &["Import 'foo' could not be resolved."],
         );
 
-        Ok(())
+        // TODO: Importing the unresolved import into a second first-party file should not trigger
+        // an additional "unresolved import" violation
+        // let b_file = system_path_to_file(&db, "/src/b.py").expect("Expected `by.py` to exist!");
+        // let b_file_diagnostics = super::check_types(&db, b_file);
+        // assert_eq!(&*b_file_diagnostics, &[]);
```

Comment:
Hmm, I'm not so sure about this and getting predictable diagnostic if we omit some might be difficult to achieve (or at least, requires a post-processing step). 

The reason why I'm unsure if that's indeed the desired behavior is that diagnostics that are a consequence of the import are missing then become harder to understand if the missing import violation only comes later in the list of diagnostics (because of how paths get sorted). One such example could be a strict mode that creates a warning whenever the type checker infers `Unknown`. 

The other reason why I think we should remove this TODO is because the elimination of the second diagnostic has to happen at a higher level than `check_types`. Calling `check_types` should return all diagnostics for that file. It's up to `workspace.check` to decide if some diagnostics should be suppressed across files.

Reply:
I'm not sure that producing a bogus diagnostic here and then filtering it out at a higher level would be appropriate. I don't think we should produce an `unresolved-import` diagnostic here at all, because there is no unresolved import here. The import is fully resolved to a symbol that exists in the symbol table of `a.py`. The fact that the symbol in `a.py` has been inferred as having type `Unknown` is immaterial to whether the import of that symbol from `a.py` into `b.py` can be resolved.

Note that this a general problem with the lint currently for any symbol that has a type inferred as `Unknown`, not just symbols that are inferred as `Unknown` because of unresolved imports specifically. This is the cause of six of the new false-positive errors in the benchmark. It's one of the problems that would be fixed if we adopted something like #12986.

For now, I'll change the comment from a TODO into an explanatory note, so that it's clear that we don't yet have agreement on where the problem should be solved

Reply:
Oh I see. I misunderstood the test case. 

I assumed that we only wanted to emit one diagnostic for:

```python
# a .py
import random # let's pretend this import doesn't exist

# b.py
import random
```